### PR TITLE
Clean up data collators and datasets

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR cleans a bit the `DataCollatorForLanguageModeling` by:
- making sure it keeps additional labels (such a next sentence labels)
- use the `tokenizer.pad` function when possible, rewrite the function that was doing the padding when it's not (to handle padding on the left as well as the right side)
- take advantage of the `special_tokens_mask` that the tokenizer can return, to avoid unnecessary conversions tensor -> list -> tensor again

As a result it deprecates `DataCollatorForSOP` (which can be replaced by `DataCollatorForLanguageModeling`) and changes the elements of `TextDatasetForNextSentencePrediction` to work with `DataCollatorForLanguageModeling`. That dataset change renders `DataCollatorForNextSentencePrediction` unusable, so it's removed (breaking change, though this is relatively contained, users that were using it should just use `DataCollatorForLanguageModeling` instead).

Tests are adapted to check `DataCollatorForLanguageModeling` does work for all those tasks.

In passing, all text datasets get their deprecation warning as we are now encouraging users to move the datasets library.

### CODE REVIEW

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -175,72 +176,111 @@ def __call__(self, features):
...
+    are_tensors_same_length = all(x.size(0) == length_of_first for x in examples)
+    if are_tensors_same_length:
+        return torch.stack(examples, dim=0)
+
+    # If yes, check if we have a `pad_token`.
+    if tokenizer._pad_token is None:
+        raise ValueError(
+            "You are attempting to pad samples but the tokenizer you are using"
+            f" ({tokenizer.__class__.__name__}) does not have one."
+        )
```
Comment (LysandreJik) [inline src/transformers/data/data_collator.py:196]:
```suggestion
            "You are attempting to pad samples but the tokenizer you are using"
            f" ({tokenizer.__class__.__name__}) does not have a pad token."
        )
```

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -175,72 +176,111 @@ def __call__(self, features):
...
 
     tokenizer: PreTrainedTokenizerBase
     mlm: bool = True
     mlm_probability: float = 0.15
 
+    def __post_init__(self):
+        if self.mlm and self.tokenizer.mask_token is None:
+            raise ValueError(
+                "This tokenizer does not have a mask token which is necessary for masked language modeling. "
+                "You should pass `mlm=False`."
```
Comment (LysandreJik) [inline src/transformers/data/data_collator.py:241]:
```suggestion
                "You should pass `mlm=False` to train on causal language modeling instead."
```

File: src/transformers/data/data_collator.py [hunk 3]
```diff
@@ -175,72 +176,111 @@ def __call__(self, features):
...
+        special_tokens_mask = batch.pop("special_tokens_mask", None)
         if self.mlm:
-            inputs, labels = self.mask_tokens(batch)
-            return {"input_ids": inputs, "labels": labels}
+            batch["input_ids"], batch["labels"] = self.mask_tokens(
+                batch["input_ids"], special_tokens_mask=special_tokens_mask
+            )
         else:
-            labels = batch.clone().detach()
+            labels = batch["input_ids"].clone().detach()
```
Comment (LysandreJik) [inline src/transformers/data/data_collator.py:260]: Out of curiosity, do we really need to `clone().detach()` the tensors? I recall having the following in the documentation:
```py
loss = model(inputs, labels=inputs)[0]
```

I guess it's probably safer to do it that way to prevent any changes done to the input IDs to impact the labels as well?

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/data/data_collator.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/data/data_collator.py:260]: It shouldn't be necessary, I agree: there is no grad attach to the inputs and the model does not change the input ids, so the clone is not necessary. Feels more like a reflex than something really useful.
